Refund: use 3rd party blockexplorer endpoints to double check if a lockup address has funds

kiocosta opened the pull request:
Closes #809
Closes #832

I have tested it this way:
1. Created the BTC->LN invoice
2. Made the HTLC expire by mining a bunch of blocks
3. Verified the swap is expired due to HTLC timeout -- not refundable yet
4. Sent sats to the lockup address
5. Verified the swap is now refundable
6. Added a refund address and refunded successfully

Did the above with L-BTC as well.

**Caveats**:
1. At this moment, only one UTXO is refundable through the web app. I will assign myself another issue to improve this.
2. It's also not possible to refund UTXOs of a successfully completed swap. Backend currently doesn't allow it: 
```
POST: http://localhost:9001/v2/swap/submarine/:swapId
Bad Request:
"{"error":"swap not eligible for a cooperative refund: status not eligible"}"
```

I believe these two edge-cases should be addressed on a different PR.

kilrau reviewed (changes requested):
Doesn't work for me on Testnet. I created a Liquid -> Lightning Swap on Testnet, didn't send any funds and set the swap status manually to `invoice.failedToPay` in our backend. Then sent funds to the lockup address. WebApp still states "NO LOCKUP TRANSACTION FOUND". I don't think the calls to the explorers are being triggered.

[Screencast from 2025-03-04 11-44-42.webm](https://github.com/user-attachments/assets/54d88bd2-5bcd-40ca-a31f-7f0e0552c632)

This can be reproduced with a regtest setup the same way, just use `boltz-cli setswapstatus sWaPiD invoice.failedToPay` in the backend.

kilrau commented:
As discussed on our call, state of the swap doesn't matter, success/failed/expired. Users send funds to lockup addresses no matter the state of a swap. Probably can exclude `swap.created` as sending funds to a lockup address of  a swap in that state is simply continuing with the swap as intended.

kilrau commented:
As discussed on our call, state of the swap doesn't matter, success/failed/expired. Users send funds to lockup addresses no matter the state of a swap. Probably can exclude swap.created as sending funds to a lockup address of a swap in that state is simply continuing with the swap.

kiocosta commented:
@kilrau I have updated the description of the PR with the changes made.
I have also re-tested it with the new changes and it should now work with `invoice.failedToPay`.

kilrau commented:
Not sure what you changed since yesterday as yesterday web app called blockstream.info/liquid/api and detected the UTXO, but Just ran latest commit `4203bee` on mainnet with the exact same Liquid submarine swap (state `invoice.failedToPay`) that has a confirmed UTXO but now it doesn't call blockstream.info and thus doesn't detect the utxo: 
![Screenshot from 2025-03-07 11-19-24](https://github.com/user-attachments/assets/55dcf0b8-d20a-45f7-b108-76225969dec2)

kilrau reviewed (changes requested):
Found the issue: calling blockstream.info works when refunding from browser storage, it does NOT work when using external refund file. That worked yesterday. Please test both scenarios. Also please test all other states apart from `invoice.failedToPay`.

kilrau commented:
Broadcasting the refund tx on liquid testnet works, on liquid mainnet not (min relay fee not met):
![image](https://github.com/user-attachments/assets/88e67d7b-ca84-4287-b18a-981befb357c5)

That sth to do with the switch to DiscountCT? If so, how can Caio make it work? @michael1011

kilrau commented:
Also Caio and I discussed that eventually we want web app to be able to handle refunds for submarine/chain swaps in *all* states, especially successful ones (`transaction.claimed`) to catch the case where users send to an old lockup address. Right now backend refuses to do this cooperatively `"{"error":"swap not eligible for a cooperative refund: status not eligible"}"`.

We can leave this for another PR, but is there anything speaking against backend re-checking a lockup address if it gets a cooperative refund request and allow cooperative refunds if there are funds in there? @michael1011

michael1011 commented:
> That sth to do with the switch to DiscountCT?

No

You are trying to refund an input that the backend does not have, it thinks that it is not swap related. Therefore, it does not allow lowball. Never has. Discount CT fixes this

michael1011 commented:
> but is there anything speaking against backend re-checking a lockup address if it gets a cooperative refund request and allow cooperative refunds if there are funds in there?

The logic to decide which things to sign cooperatively gets ***a lot more*** complicated. Let's not in this PR

kilrau commented:
> > That sth to do with the switch to DiscountCT?
> 
> No
> 
> You are trying to refund an input that the backend does not have, it thinks that it is not swap related. Therefore, it does not allow lowball. Never has. Discount CT fixes this

Ahhh got it. So once https://github.com/BoltzExchange/boltz-web-app/pull/825 is merged a simple rebase should do the trick.

kilrau commented:
> > but is there anything speaking against backend re-checking a lockup address if it gets a cooperative refund request and allow cooperative refunds if there are funds in there?
> 
> The logic to decide which things to sign cooperatively gets _**a lot more**_ complicated. Let's not in this PR

Agreed. Opened as follow up issues: https://github.com/BoltzExchange/boltz-web-app/issues/834 & https://github.com/BoltzExchange/boltz-backend/issues/839

What we should do in this PR though is to try non-cooperative refund. Like this at least it'll work even though users might have to wait for the swap to expire. And we need that fallback anyways.

kiocosta commented:
@kilrau @michael1011 
Pushed new changes. Tested refund in the following scenarios:
- using rescue key
- using old refund file
- expired HTLC
- expired invoice
- invoice paid with incorrect amount
- invoice setted to "failedToPay" through the CLI

I was able to refund the UTXO in all of the above.

From `swapStatusFailed`, the only case I don't know how to test is `transaction.failed`.
UTXOs in `transaction.refunded` or `swap.refunded` are still not refundable, as written in the beginning of the PR.

I ran all e2e tests locally. Sometimes they would break when running in parallel, but all passed when ran in isolation (not sure if this is something to be concerned or not).

kilrau commented:
> From `swapStatusFailed`, the only case I don't know how to test is `transaction.failed`.

CTRL+F in https://docs.boltz.exchange/api/lifecycle then you know when this state happens. For testing you can just set it manually on regtest backend via `boltz-cli setswapstatus sWaPiD transaction.failed`.

Once you figured out the e2e thing and CI is passing I'll do a final round of manual testing. @michael1011 an idea?

michael1011 commented on src/utils/refund.ts:
```diff
@@ -38,7 +38,7 @@ const refundTaproot = async <T extends TransactionInterface>(
     decodedAddress: DecodedAddress,
     feePerVbyte: number,
-    timeoutBlockHeight: number,
+    timeoutBlockHeight?: number,
```
We *need* that information for refunds that are not cooperative

kiocosta replied:
@michael1011 fixed

michael1011 commented on src/utils/swapCreator.ts:
```diff
@@ -20,6 +20,7 @@ import { getPair } from "./helper";
 export type SwapBase = {
     type: SwapType;
     status?: string;
+    address?: string;
```
Why add `address` here?

kiocosta replied:
@michael1011 I thought it was missing because I use it in a few places to get the UTXOs:
```
const utxos = await fetchUTXOsWithFailover(
    swap.assetSend,
    swap.address,
);
```
But this is actually only applicable to `SubmarineSwap`s, right?
I have added the typing of `SubmarineSwap` in the refund workflows used in this PR.

michael1011 commented on src/components/RefundButton.tsx:
```diff
@@ -223,19 +227,38 @@ export const RefundBtc = (props: {
+            if (utxos.length) {
+                const rawTx = await fetchRawTxWithFailover(
+                    swap.assetSend,
+                    utxos[0].txid,
```
Why only take the first one?

michael1011 commented on src/components/RefundButton.tsx:
```diff
@@ -223,19 +227,38 @@ export const RefundBtc = (props: {
+            );
 
-        return transactionToRefund;
+            if (utxos.length) {
```
```suggestion
            if (utxos.length > 0) {
```

Not a fan of relying on 0 being falsy

kiocosta replied:
@michael1011 fixed this and the other cases where this is repeated

michael1011 commented on src/pages/RefundExternal.tsx:
```diff
@@ -111,8 +112,26 @@ export const RefundBtcLike = () => {
+                        return {
+                            ...swap,
+                            transaction: {
+                                id: utxos[0].txid,
```
Why only take the first one?

michael1011 commented on src/pages/RefundExternal.tsx:
```diff
@@ -111,8 +112,26 @@ export const RefundBtcLike = () => {
+                            ...swap,
+                            transaction: {
+                                id: utxos[0].txid,
+                                vout: null,
```
The response gives us the `vout`; let's use it

kiocosta replied:
@michael1011 vout added.

michael1011 commented on src/pages/RefundExternal.tsx:
```diff
@@ -111,8 +112,26 @@ export const RefundBtcLike = () => {
             const res = await getRescuableSwaps(getXpub(source.refundJson));
-            rescueContext.setRescuableSwaps(res);
-            return res;
+            const swapsWithUTXO = await Promise.all(
+                res.map(async (swap) => {
```
Doing this for ***all** swaps in your history, will get you rate limit and/or your IP address banned from the block explorer. @kilrau I guess we should paginate this.

kiocosta replied:
@michael1011 @kilrau working on the pagination right now.

michael1011 replied:
This one needs pagination too

michael1011 commented on src/pages/Refund.tsx:
```diff
@@ -29,19 +30,32 @@ const Refund = () => {
-                    swapStatusFailed.TransactionLockupFailed === swap.status,
-            );
+        const swapsWithUTXO = (
+            await Promise.all(
+                allSwaps.filter(refundSwapsSanityFilter).map(async (swap) => {
```
Same thing here. You cannot hit the block explorer for every swap. It will get you rate limited

michael1011 commented on src/status/SwapExpired.tsx:
```diff
@@ -22,6 +26,19 @@ const SwapExpired = () => {
+            if (utxos.length) {
+                const rawTx = await fetchRawTxWithFailover(
+                    swap().assetSend,
+                    utxos[0].txid,
```
Why only the first one?

kiocosta replied:
@michael1011 currently we consider `refundTx` as a string
```
export type SwapBase = {
    ...
    refundTx?: string;
    ...
```
and use it as a condition to decide on what to render in different components: `RefundEvm`, `RefundExternal`, `SwapRefunded` and `Pay`. 

I think changing the logic of these to deal with multiple `refundTxs` may lead to more complexity in the PR, so I thought I'd do it in another PR as a way of keeping things simpler for reviewing and testing.

However, I can implement it on this PR if preferred.

michael1011 reviewed (changes requested):
> Only InvoiceFailedToPay and InvoicePending seem to be allowed for this command.

Yes. Changing the state to `transaction.failed` could lead to loss of funds in prod, which is why the command doesn't let you do it. Either change the status in the database and restart the backend or have less funds in the wallet the backend uses than it attempts to send

kiocosta commented on src/config.ts:
```diff
@@ -31,6 +31,7 @@ const defaults = {
 
 type Asset = {
     blockExplorerUrl?: Url;
+    apiUrls?: Url[];
```
@michael1011 done

kiocosta commented on src/utils/blockchain.ts:
```diff
@@ -0,0 +1,77 @@
+        block_height: number;
+        block_hash: string;
+        block_time: number;
+    };
```
@michael1011 fixed.

kiocosta commented on src/utils/blockchain.ts:
```diff
@@ -0,0 +1,77 @@
+import { chooseUrl, config } from "../config";
+
+export interface Transaction {
```
@michael1011 fixed.

kiocosta commented on src/utils/blockchain.ts:
```diff
@@ -0,0 +1,77 @@
+
+            if (!response.ok) {
+                continue;
+            }
```
@michael1011 logged with:
```
            if (!response.ok) {
                log.error(`Failed to fetch UTXOs for ${address}`);
                continue;
            }

            return (await response.json()) as UTXO[];
        } catch (e) {
            log.error(`Failed to fetch UTXOs for ${address}: ${e.stack}`);
            continue;
        }
```
I noticed `e.stack` is not commonly used on the project. I like being able to see the callstack, what do you think?
![image](https://github.com/user-attachments/assets/5a7dfecb-9047-4ad1-9ee4-37456ea7db03)

kiocosta commented on src/utils/blockchain.ts:
```diff
@@ -0,0 +1,77 @@
+
+            if (!response.ok) {
+                continue;
+            }
```
@michael1011 done like the one above.

kiocosta commented on src/utils/blockchain.ts:
```diff
@@ -0,0 +1,77 @@
+        }
+    }
+
+    throw new Error("All API endpoints failed");
```
@michael1011 updated

kiocosta commented on src/utils/boltzClient.ts:
```diff
@@ -126,6 +126,7 @@ type SubmarineCreatedResponse = {
 type ReverseCreatedResponse = {
     id: string;
     invoice: string;
+    address: string;
```
@michael1011 indeed, removed.

kiocosta commented:
@michael1011 @kilrau 

I've pushed new changes that add pagination to the refund page and address all of Michael's comments on the PR, except for those related to refunding multiple UTXOs.

I have WIP for refunding multiple UTXOs. I need more time to complete and test it, as it involves changes across multiple components.

Current pagination looks and behavior:

https://github.com/user-attachments/assets/8b16efc2-ffb2-4334-a7c3-4ef7239df7cb

@kilrau There are new translations that need review.

I chose the label "No refund due" instead of "Non-refundable" to avoid confusion. "Non-refundable" might imply that a swap can never be refunded, whereas "No refund due" clarifies that there's nothing to refund at the moment.

kilrau commented:
Looking good! :muscle: Will do more in-depth testing and push strings tomorrow.

kilrau reviewed (changes requested):
Tried all sort of scenarios (creating many swaps, close browser then send to a swap address of one of them) and web app reliably and immediately detected the swap as refundable when I reopened the page. Good!

One important edge case that should be fixed in this PR: If I refund a swap and then send funds AGAIN to the lockup address of this swap, web app correctly detects the UTXO and marks the swap as refundable BUT it doesn't reset the swap pages state to let me refund *again*:
[Screencast from 2025-03-14 12-31-10.webm](https://github.com/user-attachments/assets/8c2bf1bb-409f-49fe-9d88-0fb3a4dbc859)

Also here a list of some UI/UX issues:
1. `Refund`/`No Refund Due` Button disappears on mobile, prbly not enough space. Or did you intend that? I'd rather get rid of the date to make it fit and keep the button, the button is too important as it serves two purposes now: signify if the swap is refundable or not and well signify that one can tab on it to refund. Gotta think of sth. ![image](https://github.com/user-attachments/assets/a41daed6-c596-4649-ab30-1b5bfe905fbd)
2. Sticky pagination menu overlay when opening hamburger menu on mobile:
[Screencast from 2025-03-14 12-15-11.webm](https://github.com/user-attachments/assets/44a0d4d2-86cc-43b9-a06f-af1b3a74f577)
3. Pagination should be added as `/1`, `/2` etc slug to the url, so I can refresh in any state.

Just to say it again: I opened a [separate issue for refunding multiple UTXOs](https://github.com/BoltzExchange/boltz-web-app/issues/851), so let's take this out and focus on fixing the above and getting this PR merged.

Strings fixed. One more round :muscle:

kilrau commented:
Also pls check the failing e2e tests. Maybe try a rebase.

michael1011 commented on src/components/Pagination.tsx:
```diff
@@ -0,0 +1,132 @@
+    const { t } = useGlobalContext();
+    const [numberOfPages, setNumberOfPages] = createSignal<number[]>([]);
+
+    onMount(() => {
```
This shouldn't be done `onMount` but with `createMemo`

kiocosta replied:
@michael1011 memoized `numberOfPages` and used `onMount` just to invoke `updateDisplayedItems` on the first render.

michael1011 commented on src/components/Pagination.tsx:
```diff
@@ -0,0 +1,132 @@
+    const [numberOfPages, setNumberOfPages] = createSignal<number[]>([]);
+
+    onMount(() => {
+        const numberOfPages = [];
+        for (
+            let i = 1;
+            i <= Math.ceil(props.totalItems / props.itemsPerPage);
+            i++
+        ) {
+            numberOfPages.push(i);
+        }
```
```suggestion
const totalPages = Math.ceil(props.totalItems / props.itemsPerPage);
const numberOfPages = Array.from({ length: totalPages }, (_, index) => index + 1);
```

kiocosta replied:
@michael1011 done

michael1011 commented on src/components/Pagination.tsx:
```diff
@@ -0,0 +1,132 @@
+import "../style/pagination.scss";
+
+const Pagination = (props: {
+    items: Accessor<unknown[]>;
+    setDisplayedItems: (items: unknown[]) => void;
+    currentPage: Accessor<number>;
+    setCurrentPage: Setter<number>;
+    sort?: (items: unknown[]) => unknown[];
```
Use a generic instead of `unknown` to get type safety

kiocosta replied:
@michael1011 done

michael1011 commented on src/pages/Refund.tsx:
```diff
@@ -1,88 +1,145 @@
+import { SomeSwap, SubmarineSwap } from "../utils/swapCreator";
 import ErrorWasm from "./ErrorWasm";
 
+const SWAPS_PER_PAGE = 5;
```
```suggestion
const swapsPerPage = 15;
```

kiocosta replied:
@michael1011 done

michael1011 reviewed (commented):
Starting to look good!

kiocosta commented:
@kilrau 
>One important edge case that should be fixed in this PR: If I refund a swap and then send funds AGAIN to the lockup address of this swap, web app correctly detects the UTXO and marks the swap as refundable BUT it doesn't reset the swap pages state to let me refund again:

Fixed! Now the refund button is always available when there is any UTXO. When the user clicks the refund button, it also refunds **all** available UTXOs in that swap at that moment. I think all we have to do in #851 now is to display all the refund transaction ids, since right now I'm only displaying the last refund transaction id, but the refund of multiple UTXOs is already happening.

Fixed all the UX issues you listed as well.
For **4**, instead of adding `/1`, `/2`, I decided to use `?page=1`, `?page=2` etc, since AFAIK using the slug would require me to use [SolidJS's navigate](https://docs.solidjs.com/solid-router/reference/primitives/use-navigate) to update it. Although `navigate` updates super quick, it makes the page scroll to the top, which can be annoying to users. With query parameters I can update the list and the parameter without causing a scroll to the top. 

An invalid `?page=` value redirects user to `?page=1`.

michael1011 commented on src/pages/Refund.tsx:
```diff
@@ -1,88 +1,152 @@
+import { SomeSwap, SubmarineSwap } from "../utils/swapCreator";
 import ErrorWasm from "./ErrorWasm";
 
+const SWAPS_PER_PAGE = 10;
```
```suggestion
const swapsPerPage = 10;
```

kiocosta replied:
@michael1011 fixed!

michael1011 commented on src/utils/blockchain.ts:
```diff
@@ -0,0 +1,80 @@
+        const rawTx = await fetchRawTxWithFailover(swap.assetSend, utxo.txid);
+        txs.push({
+            hex: rawTx.hex,
+            timeoutBlockHeight: null,
```
```suggestion
            timeoutBlockHeight: swap.timeoutBlockHeight,
```

kiocosta replied:
@michael1011 this file actually had bad typing from my side. `fetchRawTxWithFailover` does not return `timeoutBlockHeight`. I have fixed the types for it tho (more on comment below).

michael1011 commented on src/utils/blockchain.ts:
```diff
@@ -0,0 +1,80 @@
+export const getSwapUTXOs = async (swap: SubmarineSwap) => {
+    const utxos = await fetchUTXOsWithFailover(swap.assetSend, swap.address);
+
+    const txs: {
+        hex: string;
+        timeoutEta?: number;
+        timeoutBlockHeight: number;
+    }[] = [];
```
Use the `LockupTransaction` from `boltzClient`

kiocosta replied:
@michael1011 the return of `fetchRawTxWithFailover` doesn't match `LockupTransaction`, `hex` is the only property in common. I have typed them this way though:
https://github.com/BoltzExchange/boltz-web-app/blob/57f86b172334bb5ff518afbb78d15828716ddd29/src/utils/blockchain.ts#L87-L90

https://github.com/BoltzExchange/boltz-web-app/blob/57f86b172334bb5ff518afbb78d15828716ddd29/src/context/Pay.tsx#L22-L24

The intention is to keep them "tied together" and accomodate the original `lockupTransaction` and `utxos` on the same array.

michael1011 commented on src/pages/RefundRescue.tsx:
```diff
@@ -49,25 +58,67 @@ const mapSwap = (
+            setSwapStatusTransaction(res.transaction);
+            setFailureReason(res.failureReason);
+
+            if (rescuableSwap().type === SwapType.Submarine) {
```
Why only for submarine swaps? Should also be done for chain swaps
Also, this does not consider swaps from RBTC

kiocosta replied:
@michael1011 fixed

michael1011 commented on src/context/Pay.tsx:
```diff
@@ -36,6 +44,9 @@ const PayProvider = (props: { children: JSX.Element }) => {
     const [swapStatus, setSwapStatus] = createSignal<string>("");
     const [swapStatusTransaction, setSwapStatusTransaction] =
         createSignal<SwapStatusTransaction>({});
+    const [refundableUTXOs, setRefundableUTXOs] = createSignal<Transaction[]>(
+        [],
+    );
```
Why is this in the context? The flow is really unclear which side effects fetches which values when. And the initial value for this is `[]` but you set it `null` in some places.

kiocosta replied:
@michael1011 I believe this should be in a context because I need the information about the UTXOs in many components, but I don't wanna prop drill it or perform duplicated requests for the block explorers.

We have three "refund flows": `Pay.tsx`, `RefundExternal.tsx` or `RefundRescue.tsx`. 
Each of these is a page which has the `RefundButton` component.

In the case of `Pay.tsx`, the `RefundButton` appear as a child of different components (`InvoiceFailedToPay`, `TransactionLockupFailed`, etc).

I need the information of the UTXOs at the page level, because I use it to conditionally render the `RefundButton`:
https://github.com/BoltzExchange/boltz-web-app/blob/2060d80ef1f8e012affa70132fdc5f5db5a77f1a/src/pages/Pay.tsx#L73-L76
https://github.com/BoltzExchange/boltz-web-app/blob/2060d80ef1f8e012affa70132fdc5f5db5a77f1a/src/pages/Pay.tsx#L114-L126

But I also need it on `RefundButton`, to perform the `refundAction` and to properly render the component.

Since the UTXO information is on the context, I don't have to make another GET request on `RefundButton` to use it. I also don't need to "prop drill" it: `Pay.tsx` -> `TransactionLockupFailed` -> `RefundButton`.

What do you think?

michael1011 replied:
Organizing this nicer would be great. Also we can surely avoid the code duplication to avoid having to maintain this code in three places

michael1011 commented on src/pages/Pay.tsx:
```diff
@@ -34,39 +35,64 @@ import TransactionClaimed from "../status/TransactionClaimed";
             setSwapStatusTransaction(res.transaction);
             setFailureReason(res.failureReason);
+
+            if (currentSwap.type === SwapType.Submarine) {
```
This is effectively duplicate code to `src/pages/RefundRescue.tsx`

michael1011 commented on src/pages/Pay.tsx:
```diff
@@ -34,39 +35,64 @@ import TransactionClaimed from "../status/TransactionClaimed";
+
+            if (currentSwap.type === SwapType.Submarine) {
+                try {
+                    const utxos = await getSwapUTXOs(
+                        currentSwap as SubmarineSwap,
+                    );
+                    setRefundableUTXOs(utxos);
```
Waiting on a block explorer API to respond to show a swap is a ***bad*** idea

michael1011 commented on src/pages/Refund.tsx:
```diff
@@ -1,88 +1,152 @@
+                            const lockupTransaction =
+                                await getLockupTransaction(swap.id, swap.type);
+
+                            const utxos = await fetchUTXOsWithFailover(
```
We ***cannot*** rely on the block explorer APIs to work

michael1011 commented on src/pages/RefundExternal.tsx:
```diff
@@ -47,20 +58,46 @@ enum RefundType {
+            return swap;
+        }
+        return null;
+    });
+
     return (
         <>
-            <Show when={refundTxId() === ""}>
+            <Show when={refundTxId() === "" && refundableSwap() !== undefined}>
```
You return null but check for undefined

kiocosta replied:
@michael1011 fixed

michael1011 commented on src/pages/RefundExternal.tsx:
```diff
@@ -47,20 +58,46 @@ enum RefundType {
+
+    const [refundableSwap] = createResource(swap, async (swap) => {
+        if (swap) {
+            if (swap.type === SwapType.Submarine) {
```
Same thing here. This also applies to chain swaps and should exclude swaps from RBTC

michael1011 commented on src/pages/RefundExternal.tsx:
```diff
@@ -112,7 +149,58 @@ export const RefundBtcLike = () => {
+        currentSwaps,
+        async (swaps: (SubmarineSwap & RescuableSwap)[]) => {
+            return await Promise.all(
+                swaps.map(async (swap) => {
+                    try {
+                        const res = await getSwapStatus(swap.id);
+
+                        if (
+                            !(await updateSwapStatus(swap.id, res.status)) &&
+                            Object.values(swapStatusFailed).includes(res.status)
+                        ) {
+                            // Make sure coins were locked for the swaps with status "swap.expired" or "swap.failedToPay"
+                            const lockupTransaction =
+                                await getLockupTransaction(swap.id, swap.type);
+
+                            const utxos = await fetchUTXOsWithFailover(
+                                swap.assetSend,
+                                swap.address,
+                            );
```
This needs pagination. Fetching transactions UTXOs for all swaps will get you banned

kiocosta replied:
@michael1011 it's already using pagination:
https://github.com/BoltzExchange/boltz-web-app/blob/2060d80ef1f8e012affa70132fdc5f5db5a77f1a/src/pages/RefundExternal.tsx#L310-L320

This `createResource` uses the `currentSwaps` set by `Pagination` and only performs requests for these:
https://github.com/BoltzExchange/boltz-web-app/blob/2060d80ef1f8e012affa70132fdc5f5db5a77f1a/src/pages/RefundExternal.tsx#L156-L160

michael1011 commented on src/pages/RefundExternal.tsx:
```diff
@@ -112,7 +149,58 @@ export const RefundBtcLike = () => {
+                                swap.address,
+                            );
+
+                            if (lockupTransaction && utxos.length > 0) {
+                                return swap;
+                            }
+
+                            if (utxos.length > 0) {
+                                return swap;
+                            }
```
We ***cannot*** rely on the block explorer API working

kilrau commented:
> One important edge case that should be fixed in this PR: If I refund a swap and then send funds AGAIN to the lockup address of this swap, web app correctly detects the UTXO and marks the swap as refundable BUT it doesn't reset the swap pages state to let me refund _again_

This works now, awesome!

> 1. `Refund`/`No Refund Due` Button disappears on mobile, prbly not enough space.

That too, looking good (even on an iPhone SE).

> 2\. Check the spacing for different languages, looks like it's proportinal to string length and just doesn't look right for German, Spanish et al

Looking good too.

> 3\. Sticky pagination menu overlay when opening hamburger menu on mobile

Fixed.

> 4\. Pagination should be added
> Instead of adding /1, /2, I decided to use ?page=1, ?page=2 etc... AFAIK using the slug would require me to use [SolidJS's navigate](https://docs.solidjs.com/solid-router/reference/primitives/use-navigate) to update it. Although navigate updates super quick, it makes the page scroll to the top, which can be annoying to users IMO. With query parameters I can update the list and the parameter without causing a scroll to the top.

Works!

> When the user clicks the refund button, it also refunds ALL available UTXOs in that swap at that moment.
> All we could do in https://github.com/BoltzExchange/boltz-web-app/issues/851 now is to display all the refund transaction ids, since right now I'm only displaying the last one, but the refund of multiple UTXOs is already happening!

As @michael1011 not in this PR to not add any fixes on topo, but this needs an urgent fix *after* merge: https://github.com/BoltzExchange/boltz-web-app/issues/855 let's tak

> Added the pagination to the `RefundExternal` page. Re-tested the refund flows multiple times in different scenarios.

Works nicely. Also tested reusing and old swap address there and was detected reliably.

kilrau commented:
> * no longer relying on 3rd party APIs to work. If they're up we use them, if not we maintain the existing behavior.

I tested this again with all scenarios I could think of with blockstream.info and liquid.network blocked:
![image](https://github.com/user-attachments/assets/23f5853f-eed4-44b0-9b21-a7d530e60be2)

Works :fire:

kilrau reviewed (approved):
> * no longer relying on 3rd party APIs to work. If they're up we use them, if not we maintain the existing behavior.

I tested this again with all scenarios I could think of with blockstream.info and liquid.network blocked. Works :fire:

kilrau commented:
Used this PR's Vercel deployment in a support ticket recently to rescue several lockup txes that web app didn't detect. Worked all as expected.
